File: src/java/voldemort/client/SocketStoreClientFactory.java
/*
 * Copyright 2008-2012 LinkedIn, Inc
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package voldemort.client;

import static voldemort.cluster.failuredetector.FailureDetectorUtils.create;

import java.net.URI;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;

import voldemort.VoldemortException;
import voldemort.client.protocol.RequestFormatType;
import voldemort.cluster.Cluster;
import voldemort.cluster.Node;
import voldemort.cluster.failuredetector.ClientStoreVerifier;
import voldemort.cluster.failuredetector.FailureDetector;
import voldemort.cluster.failuredetector.FailureDetectorConfig;
import voldemort.cluster.failuredetector.FailureDetectorListener;
import voldemort.server.RequestRoutingType;
import voldemort.store.Store;
import voldemort.store.metadata.MetadataStore;
import voldemort.store.socket.SocketDestination;
import voldemort.store.socket.SocketStoreFactory;
import voldemort.store.socket.clientrequest.ClientRequestExecutorPool;
import voldemort.store.system.SystemStoreConstants;
import voldemort.utils.ByteArray;
import voldemort.versioning.InconsistencyResolver;
import voldemort.versioning.Versioned;

/**
 * A StoreClientFactory abstracts away the connection pooling, threading, and
 * bootstrapping mechanism. It can be used to create any number of
 * {@link voldemort.client.StoreClient StoreClient} instances for different
 * stores.
 * 
 * 
 */
public class SocketStoreClientFactory extends AbstractStoreClientFactory {

    public static final String URL_SCHEME = "tcp";

    private final SocketStoreFactory storeFactory;
    private FailureDetectorListener failureDetectorListener;
    private final RequestRoutingType requestRoutingType;

    public SocketStoreClientFactory(ClientConfig config) {
        super(config);
        this.requestRoutingType = RequestRoutingType.getRequestRoutingType(RoutingTier.SERVER.equals(config.getRoutingTier()),
                                                                           false);
        this.storeFactory = new ClientRequestExecutorPool(config.getSelectors(),
                                                          config.getMaxConnectionsPerNode(),
                                                          config.getConnectionTimeout(TimeUnit.MILLISECONDS),
                                                          config.getSocketTimeout(TimeUnit.MILLISECONDS),
                                                          config.getSocketBufferSize(),
                                                          config.getSocketKeepAlive(),
                                                          config.isJmxEnabled(),
                                                          jmxId);
    }

    @Override
    public <K, V> StoreClient<K, V> getStoreClient(final String storeName,
                                                   final InconsistencyResolver<Versioned<V>> resolver) {
        if(getConfig().isLazyEnabled())
            return new LazyStoreClient<K, V>(new Callable<StoreClient<K, V>>() {

                public StoreClient<K, V> call() throws Exception {
                    return getParentStoreClient(storeName, resolver);
                }
            });

        return getParentStoreClient(storeName, resolver);
    }

    private <K, V> StoreClient<K, V> getParentStoreClient(String storeName,
                                                          InconsistencyResolver<Versioned<V>> resolver) {
        return super.getStoreClient(storeName, resolver);
    }

    @Override
    protected List<Versioned<String>> getRemoteMetadata(String key, URI url) {
        try {
            return super.getRemoteMetadata(key, url);
        } catch(VoldemortException e) {
            // Fix SNA-4227: When an error occurs during bootstrap, close the
            // socket
            SocketDestination destination = new SocketDestination(url.getHost(),
                                                                  url.getPort(),
                                                                  getRequestFormatType());
            storeFactory.close(destination);
            throw new VoldemortException(e);
        }
    }

    @Override
    protected Store<ByteArray, byte[], byte[]> getStore(String storeName,
                                                        String host,
                                                        int port,
                                                        RequestFormatType type) {
        return storeFactory.create(storeName, host, port, type, requestRoutingType);
    }

    @Override
    protected FailureDetector initFailureDetector(final ClientConfig config, Cluster cluster) {
        failureDetectorListener = new FailureDetectorListener() {

            public void nodeAvailable(Node node) {

            }

            public void nodeUnavailable(Node node) {
                if(logger.isInfoEnabled())
                    logger.info(node + " has been marked as unavailable, destroying socket pool");

                // Kill the socket pool for this node...
                SocketDestination destination = new SocketDestination(node.getHost(),
                                                                      node.getSocketPort(),
                                                                      config.getRequestFormatType());
                storeFactory.close(destination);
            }

        };

        ClientStoreVerifier storeVerifier = new ClientStoreVerifier() {

            @Override
            protected Store<ByteArray, byte[], byte[]> getStoreInternal(Node node) {
                logger.debug("Returning a new store verifier for node: " + node);
                return SocketStoreClientFactory.this.getStore(MetadataStore.METADATA_STORE_NAME,
                                                              node.getHost(),
                                                              node.getSocketPort(),
                                                              config.getRequestFormatType());
            }

        };

        FailureDetectorConfig failureDetectorConfig = new FailureDetectorConfig(config).setCluster(cluster)
                                                                                       .setStoreVerifier(storeVerifier);

        return create(failureDetectorConfig, false, failureDetectorListener);
    }

    @Override
    protected int getPort(Node node) {
        return node.getSocketPort();
    }

    @Override
    protected void validateUrl(URI url) {
        if(!URL_SCHEME.equals(url.getScheme()))
            throw new IllegalArgumentException("Illegal scheme in bootstrap URL for SocketStoreClientFactory:"
                                               + " expected '"
                                               + URL_SCHEME
                                               + "' but found '"
                                               + url.getScheme() + "'.");
    }

    @Override
    public void close() {
        this.storeFactory.close();
        if(failureDetector != null)
            this.failureDetector.removeFailureDetectorListener(failureDetectorListener);

        super.close();
    }

    public <K, V, T> Store<K, V, T> getSystemStore(String storeName,
                                                   String clusterXml,
                                                   FailureDetector fd) {
        return getRawStore(storeName,
                           null,
                           SystemStoreConstants.SYSTEM_STORE_SCHEMA,
                           clusterXml,
                           fd);
    }
}

File: bindings/cpp/src/SocketStoreClientFactory.cpp
/* -*- C++ -*-; c-basic-offset: 4; indent-tabs-mode: nil */
/*
 * Implementation for SocketStoreClientFactory class.
 * 
 * Copyright (c) 2009 Webroot Software, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

#include <voldemort/BootstrapFailureException.h>
#include <voldemort/SocketStoreClientFactory.h>
#include "SocketStore.h"
#include "RoutedStore.h"
#include "RequestFormat.h"
#include "DefaultStoreClient.h"
#include "Cluster.h"
#include "InconsistencyResolvingStore.h"
#include "TimeBasedInconsistencyResolver.h"
#include "VectorClockInconsistencyResolver.h"

#include <iostream>
#include <exception>
#include <ctype.h>

#include "boost/threadpool.hpp"

namespace Voldemort {

using namespace boost;
using namespace std;

static const string METADATA_STORE_NAME("metadata");
static const string CLUSTER_KEY("cluster.xml");
static const string STORES_KEY("stores.xml");
static const string ROLLBACK_CLUSTER_KEY("rollback.cluster.xml");

class SocketStoreClientFactoryImpl {
public:
    SocketStoreClientFactoryImpl(ClientConfig& conf);
    ~SocketStoreClientFactoryImpl();

    /** 
     * Get a raw socket store object
     */
    Store* getStore(const string& storeName,
                    const string& host,
                    int port,
                    RequestFormat::RequestFormatType type);
    
    /**
     * Retrieve the given metadata key using the bootstrap list
     */
    VersionedValue bootstrapMetadata(const string& key);

    Store* getStore(const string& storeName,
                    const string& host,
                    int port,
                    RequestFormat::RequestFormatType type,
                    bool shouldReroute);

    shared_ptr<ClientConfig> config;
    shared_ptr<ConnectionPool> connPool;
    RequestFormat::RequestFormatType requestFormatType;
    shared_ptr<threadpool::pool> threadPool;
};

SocketStoreClientFactoryImpl::SocketStoreClientFactoryImpl(ClientConfig& conf) 
    : config(new ClientConfig(conf)), connPool(new ConnectionPool(config)),
      requestFormatType(RequestFormat::PROTOCOL_BUFFERS), 
      threadPool() 
{
    //threadPool->size_controller().resize(config->getMaxThreads());
}

SocketStoreClientFactoryImpl::~SocketStoreClientFactoryImpl() {

}

Store* SocketStoreClientFactoryImpl::getStore(const string& storeName,
                                              const string& host,
                                              int port,
                                              RequestFormat::RequestFormatType type,
                                              bool shouldReroute) {
    return new SocketStore(storeName,
                           host,
                           port,
                           config,
                           connPool,
                           type,
                           shouldReroute);
}

#define THROW_BOOTSTRAP \
    throw VoldemortException("Invalid bootstrap URL " + url + \
                             ": Expected tcp://host:port");
#define EXPECT_C(char, next) \
    if (tolower(*it) != char) \
        THROW_BOOTSTRAP; \
    state = next;

static void parseBootstrapUrl(const string& url, string& host, int& port) {
    static const int STATE_T = 0;
    static const int STATE_C = 1;
    static const int STATE_P = 2;
    static const int STATE_COL1 = 3;
    static const int STATE_SL1 = 4;
    static const int STATE_SL2 = 5;
    static const int STATE_HOSTSTRING = 6;
    static const int STATE_PORTSTRING = 7;

    stringstream hostStr, portStr;
    int state = STATE_T;

    string::const_iterator it;
    for (it = url.begin(); it != url.end(); ++it) {
        switch(state) {
        case STATE_T:
            EXPECT_C('t', STATE_C);
            break;
        case STATE_C:
            EXPECT_C('c', STATE_P);
            break;
        case STATE_P:
            EXPECT_C('p', STATE_COL1);
            break;
        case STATE_COL1:
            EXPECT_C(':', STATE_SL1);
            break;
        case STATE_SL1:
            EXPECT_C('/', STATE_SL2);
            break;
        case STATE_SL2:
            EXPECT_C('/', STATE_HOSTSTRING);
            break;
        case STATE_HOSTSTRING:
            if (isalnum(*it)) 
                hostStr << *it;
            else if (*it == ':') 
                state = STATE_PORTSTRING;
            else
                THROW_BOOTSTRAP;
            break;
        case STATE_PORTSTRING:
            if (isdigit(*it)) 
                portStr << *it;
            else
                THROW_BOOTSTRAP;
            break;
        }
    }

    if (hostStr.str().length() == 0)
        THROW_BOOTSTRAP;
    if (portStr.str().length() == 0)
        THROW_BOOTSTRAP;

    host = hostStr.str();
    portStr >> port;
    if (port == 0)
        THROW_BOOTSTRAP;
}

VersionedValue SocketStoreClientFactoryImpl::bootstrapMetadata(const string& key) {
    std::list<std::string>* boots = config->getBootstrapUrls();
    std::list<std::string>::const_iterator it;
    for (it = boots->begin(); it != boots->end(); ++it) {
        try {
            string host;
            int port;
            parseBootstrapUrl(*it, host, port);

            auto_ptr<Store> store(getStore(METADATA_STORE_NAME,
                                           host,
                                           port,
                                           requestFormatType,
                                           false));

            auto_ptr<std::list<VersionedValue> > vvs(store->get(key));
            if (vvs->size() == 1) {
                return vvs->front();
            }
        } catch (std::exception& e) {
            cerr << "Warning: Could not bootstrap '" << *it << "': "
                 << e.what() << endl;
        }
    }
    throw BootstrapFailureException("No available bootstrap servers found!");
}

SocketStoreClientFactory::SocketStoreClientFactory(ClientConfig& conf) {
    pimpl_ = new SocketStoreClientFactoryImpl(conf);

}

SocketStoreClientFactory::~SocketStoreClientFactory() {
    if (pimpl_) 
        delete pimpl_;
}

StoreClient* SocketStoreClientFactory::getStoreClient(std::string& storeName) {
    shared_ptr<InconsistencyResolver> nullResolver;
    return getStoreClient(storeName, nullResolver);
}

StoreClient* SocketStoreClientFactory::
getStoreClient(std::string& storeName,
               shared_ptr<InconsistencyResolver>& resolver) {
    shared_ptr<Store> store(getRawStore(storeName, resolver));
    return new DefaultStoreClient(store,
                                  pimpl_->config);

}

Store* SocketStoreClientFactory::getRawStore(std::string& storeName,
                                             shared_ptr<InconsistencyResolver>& resolver) {
    VersionedValue clustervv = pimpl_->bootstrapMetadata(CLUSTER_KEY);
    const std::string* clusterXml = clustervv.getValue();
    shared_ptr<Cluster> cluster(new Cluster(*clusterXml));

    shared_ptr<std::map<int, shared_ptr<Store> > > 
        clusterMap(new std::map<int, shared_ptr<Store> >());

    const std::map<int, boost::shared_ptr<Node> >* nodeMap = cluster->getNodeMap();
    std::map<int, boost::shared_ptr<Node> >::const_iterator it;
    for (it = nodeMap->begin(); it != nodeMap->end(); ++it) {
        shared_ptr<Store> store(pimpl_->getStore(storeName,
                                                 it->second->getHost(),
                                                 it->second->getSocketPort(),
                                                 pimpl_->requestFormatType,
                                                 true));
        (*clusterMap)[it->second->getId()] = store;
    }

    //VersionedValue storevv = pimpl_->bootstrapMetadata(STORES_KEY);
    //const std::string* storesXml = storevv.getValue();

    shared_ptr<Store> routedStore(new RoutedStore(storeName,
                                                  pimpl_->config,
                                                  cluster,
                                                  clusterMap,
                                                  pimpl_->threadPool));

    /* XXX - TODO Add inconsistency resolver */
    InconsistencyResolvingStore* conStore = new InconsistencyResolvingStore(routedStore);
    try {
        shared_ptr<InconsistencyResolver> 
            vcResolver(new VectorClockInconsistencyResolver());
        conStore->addResolver(vcResolver);
        
        if (resolver.get()) {
            conStore->addResolver(resolver);
        } else {
            shared_ptr<InconsistencyResolver> 
                tbResolver(new TimeBasedInconsistencyResolver());
            conStore->addResolver(tbResolver);
        }
    } catch (...) {
        if (conStore) delete conStore;
        throw;
    }

    return conStore;
}

} /* namespace Voldemort */
